### PR TITLE
Add --ftime-trace to dmd

### PR DESCRIPTION
**Pull request by dkorpel**
Reboot of https://github.com/dlang/dmd/pull/13965

This feature has been in LDC for a while now with success, and inclusion into dmd has been pre-approved a long time ago by Walter, but the PR has stalled since then. There's improvements possible with choosing better zones and refactoring the code, but for now I focused on making it closely match LDC's current implementation. 

For this PR, I still need to include the `timetrace2txt` tool, and test it a bit better.

### CODE REVIEW

File: compiler/src/dmd/timetrace.d
Comment (ibuclaw) [inline compiler/src/dmd/timetrace.d:125]: These are defined but never used.  Though actually using these would be preferable to a D struct, as it at least would be overridable to allow alternative time tracking engines to be used.

File: compiler/src/dmd/timetrace.d
Comment (ibuclaw) [inline compiler/src/dmd/timetrace.d:91]: Consider returning an OutBuffer, or moving this `buf` local to a `ref` parameter.  Doing IO inside public API is regarded as bad style, and rules it out from being unittest-able.

File: compiler/src/dmd/timetrace.d
Comment (dkorpel) [reply compiler/src/dmd/timetrace.d:125]: As mentioned in the PR description, it's currently closely matching LDC's code. I have lots of ideas to refactor the code to be better DMD style, but I'd like to know from LDC maintainers if they plan on keeping their own copy of timetrace.d or whether they want to share with upstream. In that case we need to keep a C++ interface.

Comment (ibuclaw) [reply compiler/src/dmd/timetrace.d:125]: > In that case we need to keep a C++ interface.

That is what the second sentence is advocating for. The frontend semantic passes are using a D interface and RAII. It would integrate better if it used these C++ hooks instead.

(It would be even better if each stage being traced had an enum value, rather than passing a string name and detail, but one request at a time to nudge it in the right direction).

File: compiler/src/dmd/timetrace.d
Comment (maxhaton) [reply compiler/src/dmd/timetrace.d:91]: good point

File: compiler/src/dmd/timetrace.d
Comment (JohanEngelen) [reply compiler/src/dmd/timetrace.d:125]: In LDC, we call these C++ functions in a RAII class and only the RAII class is ever used. It is very bug prone to _not_ use RAII, so the D code also uses a RAII struct.

File: changelog/dmd.ftime-trace.dd
Comment (JohanEngelen) [inline changelog/dmd.ftime-trace.dd]: I think the timetrace2txt tool is a necessity for practical use. Traces get very large quite quickly, and then the browser-based visual tooling becomes pretty slow (last time I looked). `tracy` https://github.com/wolfpld/tracy does have good performance on large traces though.

Comment (dkorpel) [reply changelog/dmd.ftime-trace.dd]: Yes, it probably needs to be added to https://github.com/dlang/tools and https://github.com/dlang/installer. My plan is to get this PR approved first and then follow up by adding timetrace2txt, since this PR is already pretty big.

File: compiler/src/dmd/timetrace.d
Comment (dkorpel) [reply compiler/src/dmd/timetrace.d]: I moved file IO to main.d

File: compiler/src/dmd/timetrace.d
Comment (dkorpel) [reply compiler/src/dmd/timetrace.d]: I added an enum value for each stage, and now use D functions (because they take a `delegate`) with help of `scope (exit)` for safety, while also keeping the C++ interface for LDC.

File: compiler/src/dmd/timetrace.d
Comment (JohanEngelen) [inline compiler/src/dmd/timetrace.d:10]: add yourself here :)

File: compiler/src/dmd/timetrace.d
Comment (JohanEngelen) [inline compiler/src/dmd/timetrace.d:136]: `extern(C++)` so it can be used by LDC.

File: compiler/src/dmd/timetrace.d
Comment (JohanEngelen) [inline compiler/src/dmd/timetrace.d:117]: TimeTraceEventType.generic as (default) parameter

**Comment by dkorpel** [comment]
I've tested what the overhead is of normal compilation without `-ftime-trace` on a copy of Phobos.

```
hyperfine "dmds0 -i -c mystd/package.d" "dmds1 -i -c mystd/package.d"
Benchmark 1: dmds0 -i -c mystd/package.d
  Time (mean ± σ):      1.540 s ±  0.019 s    [User: 1.213 s, System: 0.316 s]
  Range (min … max):    1.520 s …  1.576 s    10 runs
 
Benchmark 2: dmds1 -i -c mystd/package.d
  Time (mean ± σ):      1.553 s ±  0.011 s    [User: 1.206 s, System: 0.335 s]
  Range (min … max):    1.538 s …  1.572 s    10 runs
 
Summary
  dmds0 -i -c mystd/package.d ran
    1.01 ± 0.01 times faster than dmds1 -i -c mystd/package.d
```

(`dmds0` = master branch, `dmds1` = ftime-trace branch)

So ~1% overhead, which is worth it when you see how much build time this feature can save.

File: compiler/src/dmd/timetrace.d
Comment (JohanEngelen) [reply compiler/src/dmd/timetrace.d:136]: overall: think a little about what interface to expose to C++.  Delegates are not possible, but it'd be good if C++ can use the TimeTraceEventType, to have output be nicely uniform across backends, and also provide extra string input for that event.

File: compiler/src/dmd/timetrace.d
Comment (ibuclaw) [inline compiler/src/dmd/timetrace.d:144]:
```suggestion
    if (timeTraceProfilerEnabled)
    {
      version (IN_GCC)
        timeTraceEndEvent(eventType);
      else
        timeTraceProfiler.endScope(
```
Something like this, so long as the function is `extern(C++)`.

File: compiler/src/dmd/timetrace.d
Comment (ibuclaw) [inline compiler/src/dmd/timetrace.d:119]:
```suggestion
    if (timeTraceProfilerEnabled)
      version (IN_GCC)
        timeTraceBeginEvent(eventType);
      else
        timeTraceProfiler.beginScope(null, null, Loc.initial);
```
Likewise it should be `extern(C++)`.

Saying that if the API is fine, then I could just provide my own implementation of this module that contains none of the private parts.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@ibuclaw @JohanEngelen anything else blocking this PR?

**Pull request merged**
